### PR TITLE
Task#4

### LINKED ISSUE #29: Zadanie 4 (dla ambitnych, klasa  Time)
Spróbuj napisać klasę Time, która będzie odpowiadać za zarządzanie czasem w grze.

Klasa ta powinna informować inne klasy, takie jak Cargo, Ship, Stock o upłynięciu każdego dnia.

Poczytaj czym jest wzorzec projektowy Observer.

### PR DESCRIPTION
**Pull request by Vicx95**
resolves #29

### CODE REVIEW

File: shm/inc/delegate.hpp [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+#include <iostream>
+class Delegate {
+public:
+    virtual void payCrew(size_t money) = 0;
+    virtual ~Delegate() {}
+};
```
Comment (arafalski) [inline shm/inc/delegate.hpp]: New line

File: shm/inc/time.hpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    Time();
+    void addObserver(Observer* obs);
+    void removeObserver(Observer* obs);
+    size_t getElapsedTime() const;
+    Time& operator++();
+
+private:
+    size_t timeElapsed_;
+    std::vector<Observer*> observers_;
+};
```
Comment (arafalski) [inline shm/inc/time.hpp]: New line

File: shm/src/globaltime.cpp [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#include "globaltime.hpp"
+
+Time* GlobalTime::globalTime_ = nullptr;
```
Comment (arafalski) [inline shm/src/globaltime.cpp]: New line

File: shm/src/ship.cpp [hunk 4]
```diff
@@ -102,3 +109,7 @@ void Ship::receiveCargo(Cargo* cargo, size_t amount, CargoHolder* cargoHolder) {
 void Ship::clearEmptyCargos() {
     cargo_.erase(std::remove_if(cargo_.begin(), cargo_.end(), [](auto& cargo) { return cargo->getAmount() == 0; }), cargo_.end());
 }
+
+Ship::~Ship(){
+    time_->removeObserver(this);
+}
```
Comment (arafalski) [inline shm/src/ship.cpp]: New line

File: shm/src/alcohol.cpp [hunk 5]
```diff
@@ -12,6 +15,15 @@ size_t Alcohol::getPrice() const {
     return static_cast<size_t>(static_cast<float>(basePrice_) * (power_ / kBasePower));
 }
 
+
```
Comment (arafalski) [inline shm/src/alcohol.cpp]: Delete this line

File: shm/inc/delegate.hpp [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+#include <iostream>
+class Delegate {
```
Comment (arafalski) [inline shm/inc/delegate.hpp]:
```suggestion
#pragma once

#include <iostream>

class Delegate {
```

File: shm/src/time.cpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    observers_.erase(std::remove(observers_.begin(), observers_.end(), obs), observers_.end());
+}
+
+size_t Time::getElapsedTime() const {
+    return timeElapsed_;
+}
+
+Time& Time::operator++() {
+    timeElapsed_++;
+    for (const auto observer : observers_) {
```
Comment (kontrybutor) [inline shm/src/time.cpp]:
```suggestion
    for (const auto& observer : observers_) {
```

**Review by kontrybutor — APPROVED** [review]

**Review by arafalski — APPROVED** [review]

**Pull request merged**